rosbag2_cpp/include/rosbag2_cpp/cache/message_cache_circular_buffer.hpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_CPP__CACHE__MESSAGE_CACHE_CIRCULAR_BUFFER_HPP_
+#define ROSBAG2_CPP__CACHE__MESSAGE_CACHE_CIRCULAR_BUFFER_HPP_
+
+#include <memory>
+#include <list>
+
+#include "rosbag2_cpp/visibility_control.hpp"
+#include "rosbag2_storage/serialized_bag_message.hpp"
+
+// This is necessary because of using stl types here. It is completely safe, because
+// a) the member is not accessible from the outside
+// b) there are no inline functions.
+#ifdef _WIN32
+# pragma warning(push)
+# pragma warning(disable:4251)
+#endif
+
+namespace rosbag2_cpp
+{
+namespace cache
+{
+
+class ROSBAG2_CPP_PUBLIC MessageCacheCircularBuffer
+{
+public:
+  explicit MessageCacheCircularBuffer(const uint64_t max_cache_size);
+
+  using buffer_element_t = std::shared_ptr<const rosbag2_storage::SerializedBagMessage>;
+
+  /**
+  * If buffer size has some space left, we push the message regardless of its size,
+  *  but if this results in exceeding buffer size, we begin dropping old messages.
+  */
+  void push(buffer_element_t msg);
+
+  /// Clear buffer
+  void clear();
+
+  /// Get number of elements in the buffer
+  size_t size();
+
+  /// Get buffer data
+  const std::list<buffer_element_t> & data();
+
+private:
+  std::list<buffer_element_t> buffer_;
+  uint64_t buffer_bytes_size_ {0u};
+  const uint64_t max_bytes_size_;
```

Comment:
In case when will be used default constructor `max_bytes_size_` will be uninitialized. It will lead to the undefined behavior for further usage of default constructed instance of the class.
Since it's declared as const will need to explicitly define default constructor without parameters and initialize it to zero.

Reply:
You could explicitly delete the default constructor to avoid this concern.

```
MessageCacheCircularBuffer() = delete;
```

Reply:
Going to go with deleting the default constructor since we'll require a non-zero `max-cache-size` to be provided by the user whenever snapshot mode is enabled.

Reply:
Ok guys, if you want to delete default constructor, how are you going to use this class with `shared_ptr`?
As far as I am understand the code is not going to compile in the way how it's uses.

Reply:
Currently we're constructing each `std::shared_ptr` for MessageCacheCircularBuffer via `std::make_shared` and explicitly passing the `max-cache-size` as a parameter, so I don't think deleting the default constructor is an issue here. From testing, it also compiles without issue and passes all test cases.

Reply:
Ok, then.